code/datums/components/object_possession.dm

```diff
@@ -0,0 +1,115 @@
+/// Component that allows admins to control any object as if it were a mob.
+/datum/component/object_possession
+	/// Stores a reference to the mob that is currently possessing the object.
+	var/datum/weakref/poltergeist = null
+	/// Ref to the screen object that is currently being displayed.
+	var/datum/weakref/screen_alert_ref = null
+	/**
+	  * back up of the real name during admin possession
+	  *
+	  * If an admin possesses an object it's real name is set to the admin name and this
+	  * stores whatever the real name was previously. When possession ends, the real name
+	  * is reset to this value
+	  */
+	var/stashed_name = null
+	/// List of signals we register to in order to know when to end possession.
+	var/static/list/signals_to_delete_on = list(
+		COMSIG_END_OBJECT_POSSESSION_VIA_COMPONENT_CHAIN,
+		COMSIG_END_OBJECT_POSSESSION_VIA_SCREEN_ALERT,
+		COMSIG_END_OBJECT_POSSESSION_VIA_VERB,
+	)
+
+
+/datum/component/object_possession/Initialize(mob/user)
+	. = ..()
+	if(!isobj(parent) || !ismob(user))
+		return COMPONENT_INCOMPATIBLE
+
+	if(HAS_TRAIT(user, TRAIT_CURRENTLY_CONTROLLING_OBJECT))
+		SEND_SIGNAL(user, COMSIG_END_OBJECT_POSSESSION_VIA_COMPONENT_CHAIN) // end the previous possession before we start the next one
+
+	var/obj/obj_parent = parent
+
+	if((obj_parent.obj_flags & DANGEROUS_POSSESSION) && CONFIG_GET(flag/forbid_singulo_possession))
+		to_chat(user, "[obj_parent] is too powerful for you to possess.", confidential = TRUE)
+		return COMPONENT_INCOMPATIBLE
+
+	if(HAS_TRAIT(user, TRAIT_CURRENTLY_CONTROLLING_OBJECT))
+		stashed_name = user.name
+
+	ADD_TRAIT(user, TRAIT_CURRENTLY_CONTROLLING_OBJECT, REF(parent))
+
+
+	stashed_name = user.real_name
+	poltergeist = WEAKREF(user)
```

Comment:
Since we register signals onto poltergeist / user, maybe this should be a real managed reference rather than a weakref? We have a vested interest in poltergist existing so we probably care about it

Reply:
rule of thumb for weakrefs is to only hold strong references we _own_. this component does not own the mob in that thought process, we care about the mob existing but we only loosely need it.

probably do need to attach a `on_user_qdel()` proc though

Reply:
Does the component have a reason to exist if the controller is deleted? Does the object stay possessed even though it has no controller? If not then we should probably "own" it